### ISSUE TITLE
Missing features

### ISSUE THREAD

**Issue by LearnWithTung** [mapbox/mapbox-maps-ios#991]
## New Feature
I am having trouble understanding Mapbox v10. The way I'm using it now it seems to be missing many functions like the regionDidChangeAnimated for example that was a part of the MGLMapViewDelegate prior to v10.
The way I'm interacting with the map now is through `mapView.mapboxMap.onEvery()` but the map event options here don't include equivalents to many of the old delegated functions.

View annotations is also a thing, I can not filter out annotations as before like bellow
```
map.annotations?.filter {...}
```

Thank you all in advance for any help you can provide.

**Comment by ZiZasaurus** [comment]
@LearnWithTung, can you share more about your use case for the delegate methods?  

Currently, in order to filter view annotations, you would have to manually adjust the ViewAnnotationOptions.visible for the annotations that you want to show/hide.

**Comment by LearnWithTung** [comment]
@ZiZasaurus 
In my current implementation, we use `regionDidChangeAnimated` combine with `MGLCameraChangeReason` to specify if the region is changed by user.
We also use ` regionIsChanging` and some other delegate methods but cannot find equivalent methods in the new SDK

**Comment by ZiZasaurus** [comment]
@LearnWithTung, can you please provide a code snippet of your implementation prior to v10 so I can point you in the right direction?

**Comment by tungvustarglobal** [comment]
@ZiZasaurus 
We're using `MGLMapViewDelegate` and RxSwift to convert delegate methods to observables, here's a code snippet
```
var regionWillChange: ControlEvent<(MGLMapView, Bool)> {
    let selector = #selector(MGLMapViewDelegate.mapView(_:regionWillChangeAnimated:))
    let source = delegate.methodInvoked(selector).map { args in
      return (try Reactive.castOrThrow(MGLMapView.self, args[0]),
      try Reactive.castOrThrow(Bool.self, args[1]))
    }
    return ControlEvent(events: source)
  }
  
  var regionDidChange: ControlEvent<(MGLMapView, Bool, Bool)> {
    let selector = #selector(MGLMapViewDelegate.mapView(_:regionDidChangeWith:animated:))
    let source = delegate.methodInvoked(selector).map { args -> (MGLMapView, Bool, Bool) in
      let reason = try MGLCameraChangeReason(rawValue: Reactive.castOrThrow(UInt.self, args[1]))
      return (try Reactive.castOrThrow(MGLMapView.self, args[0]),
              reason.contains(.programmatic),
              try Reactive.castOrThrow(Bool.self, args[2]))
    }
    return ControlEvent(events: source)
  }
  ```
  
  I also want to share you APIs that we're currently use to update visibility of layers on the map but can not find in the new framework (filter annotations and predicate)
  ```
  func updateBayPointsVisibility() {
    let map = rootView.mapView!
    
    let carParkAnnotations = (map.annotations?.filter { $0 is SomeClass }) as? [SomeClass]
    let visibleCarParkIDs = Set(carParkAnnotations?.map { $0.id } ?? [])
    
    // for all layers, which display car park icon apply new visibility predicate
    [
      GlobalConstants.Layers.layer1ID,
      GlobalConstants.Layers.layer2ID,
      GlobalConstants.Layers.layer3ID,
      GlobalConstants.Layers.layer4ID
    ].forEach { layerID in
      let bayIcon = map.style?.layer(withIdentifier: layerID) as? MGLVectorStyleLayer
      bayIcon?.predicate = // predicate logic
    }
  }

**Comment by irlanco** [comment]
I agree that `regionDidChangeAnimated` is a very important delegate/feature that seems to be missing in v10. We use it to repopulate annotations when the viewport changes. 

Do the devs have a suggestion to mimic this behavior? Basically just knowing when the user has moved the map.

**Comment by LearnWithTung** [comment]
@ZiZasaurus regarding view annotations, is there anyway to fetch all the rendered view based annotations in the map
I see that the `ViewAnnotationManager` only has `add` and `remove` with a specific view

**Comment by teradyl** [comment]
Found it, looks like it's the `GestureManagerDelegate`

https://docs.mapbox.com/ios/maps/guides/user-interaction/

**Comment by sobri909** [comment]
I'm also struggling with missing features in `ViewAnnotationManager`. In my case it's not being able to get ids for views, or to be able to walk the annotations and have visibility of all three details: view, options, and id. View and options are exposed in the `annotations` dictionary property, but ids are not.

I have a method from my pre-v10 code for `clearTheMap(except: [UUID])`. I can make this work for other Annotation types, but for view annotations I'm having to walk the `except` array and get views individually by id, then `removeAll()` of the view annotations on the map, then add back in the ones that matched the `except`. This creates visual artefacts with broken animations, etc, and isn't shippable.

It should be possible to either do `id(for: UIView)` or... well I don't know. The current API feels awkward, but I don't off the top of my head have great ideas how to fix it. Either way, as it stands now I'm going to have to maintain references to the views and ids external of Mapbox.

**Comment by sobri909** [comment]
I've come to this solution. Someone please tell me I'm being silly and missing something much more sensible: 

```swift
if let viewAnnotations = map?.viewAnnotations {
    let toKeep = except.compactMap { viewAnnotations.view(forId: $0.uuidString) }
    let toRemove = viewAnnotations.annotations.keys.filter { !toKeep.contains($0) }
    toRemove.forEach { viewAnnotations.remove($0) }
}
```
